repair/row_level.cc

```diff
@@ -650,18 +653,59 @@ future<std::list<repair_row>> to_repair_rows_list(repair_rows_on_wire rows, sche
     co_return std::move(row_list);
 }
 
+std::unordered_set<dht::token> get_tokens_from_repair_row_list(const std::list<repair_row>& rows) {
+    std::unordered_set<dht::token> ret;
+    for (auto& row : rows) {
+        // Get the tokens for the rows that are read from the node itself and
+        // ignore the rows received from peer nodes. A row with dirty_on_master
+        // flag set comes from peer node. See to_repair_rows_list.
+        if (!row.dirty_on_master()) {
+            ret.insert(row.get_dk_with_hash()->dk.token());
+        }
+    }
+    return ret;
+}
+
+// Check if the row is expired and it does not shadow any data read from the local node.
+bool can_skip_expired_mf(schema_ptr s, const dht::token& t, mutation_fragment& mf,
```

Comment:
The expired data logically become tombstones once expired and so they are subject to the same purging rules.
Why do we treat expired data differently than a tombstone?

Reply:
The logic in this patch does not change the way we purge tombstones. It is what it is now. Nothing changes.  It only changes if  a node should receive expired rows or not from a peer node.  It the expired data shadows nothing locally, there is no point to receive the expired data from the peer node. This seems pretty safe.

Reply:
> The logic in this patch does change how we purge tombstones because the method below does not take purge-ability into consideration, it only looks at whether the tombstone covers anything locally.
This in effect changes the purge strategy from gc grace to repair.

Are you saying this patch changes the tombstone_gc::repair mode in effect?  That is not the case.  How to purge tombstone is not changed. It is still the business of compaction which consults either gc_grace_seconds or repair history.

Note, the rows we could possibly skip to write in this patch come from the peer nodes only.  I am trying to find a case which  is not safe to skip the expired rows. So far, I could not find one. 

> Also, this logic is incomplete: for example it will still replicate expired rows, if there is even a single live cell in the row.

Yes, as a starter, the granularity is whole row. It is not at the level of cell. Skipping at the granularity of cell is more precise but needs more work and more complicated.

Reply:
I'll try to rephrase https://github.com/scylladb/scylladb/pull/14148#discussion_r1218986084:

Once data expires it becomes a tombstone, semantically, and compaction may turn it into an actual tombstone, losing track of the original data that expired.

If we are willing to skip expired rows if they don't shadow local data on the receiver, why not consider that for tombstones using the same condition?

Reply:
I did consider handling the tombstones in a similar way. In theory, we could drop the tombstones on the receiver if they do not shadow any local data.

I am picking expired rows only in this patch because

1) It is the easiest but with high gain. We could drop the whole row including all the cells with a very simple check.
I think expiring the whole row is a more common use case.

2) In practice, we have a high chance if ttl is used, all data will be dropped using ttl. No delete will be used. So, even if we somehow messed up something, skipping the expired rows won't be a problem since whoever has the writes will expire the data automatically anyway. No data resurrection could happen.

3) Skipping tombstones on the receiver will be more aggressive and more dangerous and more complicated.

We have tombstones for partition / row / cell and range tombstones.